docs/key-issues/multiple_architectures.md

```diff
@@ -0,0 +1,350 @@
+      deferred to the person downloading the wheels (i.e., the app builder).
+      This can be problematic as it may require expert knowledge about the
+      package being merged (such as optional modules or header files that may
+      not be present in both thin artefacts). Universal2 artefacts captures
+      this knowledge by requiring the maintainers of the wheel-producing
+      project to resolve any merging issues.
+
+    Arguments against:
+
+    - `universal2` wheels are never necessary for users installing into a
+      locally installed Python environment exclusively for their own use,
+      which is the default experience most users have with Python.
+    - The tradeoff between download size and disk space usage vs. the upside
+      for say a .dmg installer is bad - for a typical PyData stack it takes
+      hundreds of MBs per Python environment more than thin wheels, and users
+      are likely to have quite a few environments on their system at once.
+      Meaning that defaulting to `universal2` would use several GBs of disk
+      space more.
+
+        - Disk space on older MacBook Air models is 128 GB, and up to half of that
+          can be taken up by the OS and system data itself. So a few GBs can be
+          significant.
+        - Internet plans in many countries are not unlimited; almost doubling the
+          download size of wheels is a serious cost, and not desirable for any
+          user - but especially unfriendly to users in countries where network
+          infrastructure is less developed.
+
+    - In addition, it takes extra space on PyPI (examples: `universal2` wheels
+      cost an extra 81.5 MB for NumPy 1.21.4 and 175.5 MB for SciPy 1.9.1), and
+      projects with large wheels often run into total size limits on PyPI.
+    - It imposes an extra maintenance burden for each project, because separate
+      CI jobs are needed to build and test `universal2` wheels. Typically
+      projects make tradeoffs there, because they cannot support every
+      platform. And `universal2` doesn't meet the bar for usage frequency /
+      user demand here - it's well below the demand for `musllinux`,
+      `ppc64le`, PyPy, and other such platforms with still patchy support
+      (see [Expectations that projects provide ever more wheels](../../meta-topics/user_expectations_wheels.md)
+      for more on that).
+    - When a project provides thin wheels (which is a must-do for projects with
+      native code, because those are the better experience due to smaller
+      size), you cannot even install a `universal2` wheel with pip from PyPI at
+      all. Why upload artifacts you cannot install?
+    - It is straightforward to fuse two thin wheels with `delocate-fuse` (a
+      tool that comes with [delocate](https://pypi.org/project/delocate/)),
+      it's a one-liner: `delocate-fuse $x86-64_wheel $arm64_wheel -w .`
+      However, it's worth noting that this requires that any headers or python
```

Comment:
I don't think this edit is right. The exact same bug will still be present in universal2 wheels. Anything that depends on the build machine's architecture is going to be incorrect and must be explicitly handled for universal2 I believe. There may be exceptions, but I think it's true in general (similar issues as when cross-compiling) and it was true for the `long double` size issue that Isuru linked.

The problem here is the existence of universal2 as a format more than how you produced it.

Reply:
Actually, it's a more general problem even for any multi-architecture support. Typical header generation that does things like `sizeof(<C type name>)` is going to be incorrect.

Reply:
Also - doesn't this *support* my point? Building a fat wheel *isn't* a "straightforward" matter of running `delocate-fuse`".

Reply:
Sure, anything is possible, but it requires care. 

> Building a fat wheel _isn't_ a "straightforward" matter of running `delocate-fuse`".

I think that in the case of no bugs like architecture-dependent headers, it _is_ straightforward. And if there are such bugs, I'd rather get a loud clash from `delocate-fuse` rather than a silent bug.

In the end, I think the most robust process would be to build thin apps based on only thin wheels, and then only fuse the final thin apps together. That way, most/all bugs like this are likely irrelevant.

Reply:
Unfortunately, building thin apps isn't a viable option. It causes a degraded and user experience on macOS, and fat apps are a *requirement* for iOS/Android distribution.

Reply:
(noting that the "fat" interpretation on Android is a little easier to accomodate because of how it handles binaries)

Reply:
> Unfortunately, building thin apps isn't a viable option. It causes a degraded and user experience on macOS, and fat apps are a _requirement_ for iOS/Android distribution.

This is why I said "and then only fuse the final thin apps together". I meant "into fat aoos to distribute to users". The end result is the same, a fat app. It's just that the process is better controlled; any intermediate header issues get avoided when you use a stack of thin wheels.

Reply:
Your edit notes that delocate-fuse might succeed but produce broken artefacts.

But it's not an inherent problem with delocate-fuse, it's a "simple" matter of the project not supporting or testing universal2. If a project supports or tests universal2, it should not matter whether the project personally executes the compiler in multi-pass mode, or personally executes delocate-fuse.

Either both work, or both do not work.

And if/when the delocate-fuse tooling is improved, then it *does* become a straightforward matter of "attempt to fuse the wheels, if it doesn't work then submit a bug report to the project". No more "the project doesn't test this, so there may be subtle runtime bugs".

Reply:
I have rephrased this now to say: 

_Note though that robustness improvements in delocate-fuse for more complex cases (e.g., generated header files with architecture-dependent content) are needed (see [delocate#180](https://github.com/matthew-brett/delocate/issues/180)). Such cases are likely to be equally problematic for direct universal2 wheel builds (see, e.g., [numpy#22805](https://github.com/numpy/numpy/pull/22805))._

I think that addressed it, please resolve if this looks good.